File: three_number_sum/three_number_sum.py
def three_number_sum(arr, targetsum):
    output = []

    # Sorting the array
    arr.sort();

    # len -2 is used to ensure we always have two number ie. left & right pointers
    for i in range(len(arr) - 2):
        # Initialize left and right pointer with positions of array
        lptr = i + 1
        rptr = len(arr) - 1

        # Loop to control the moving of left and right positions
        while lptr < rptr:
            current_sum = arr[i] + arr[lptr] + arr[rptr]

            # CHECKPOINT 1
            if current_sum == targetsum:
                output.append([arr[i], arr[lptr], arr[rptr]])
                
                # Moving both the pointer positions one step
                lptr = lptr + 1
                rptr = rptr - 1

            # CHECKPOINT 2
            if current_sum < targetsum:
                lptr = lptr + 1

            # CHECKPOINT 3
            if current_sum > targetsum:
                rptr = rptr - 1

    return output






# EXECUTION BEGINS HERE
if __name__ == "__main__":
    # Inputs
    arr = [12, 3, 1, 2, -6, 5, -8, 6]
    targetsum = 0

    # Output: 
    # Find all possible three number combination which adds to
    # the targetsum value
    output = three_number_sum(arr, targetsum);
    print(output)

File: three_number_sum/three_number_sum.cpp
#include <iostream>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

void three_number_sum(vector<int>& arr, int targetsum, vector< vector<int>>& out) {
   // Sort the array
   sort(arr.begin(), arr.end());

   for (int i = 0; i < arr.size() -2; i++) {
      // Initialize right & left pointer with positional values
      int lptr = i + 1;
      int rptr = arr.size() - 1;

      // Loop to move the left & right pointers
      while (lptr < rptr) {
         int current_sum = arr[i] + arr[lptr] + arr[rptr];

         // CHECK 01
         if (current_sum == targetsum) {
            // Adding the current value in to output vector
            vector<int> tmp = {arr[i], arr[lptr], arr[rptr]};
            out.push_back(tmp);

            // Moving both the pointers
            lptr++;
            rptr--;
         }

         // CHECK 02
         if (current_sum < targetsum)
            lptr++;

         // CHECK 03
         if (current_sum > targetsum)
            rptr--;
      }
   }
}





// EXECUTION BEGINS HERE
int main() {
   // INPUTS
   vector<int> arr = {12, 3, 1, 2, -6, 5, -8, 6};
   int targetsum = 0;

   // OUTPUTS [[-8, 2, 6], [-8, 3,5], [-6, 1, 5]]
   vector< vector<int>> out;

   three_number_sum(arr, targetsum, out);

   // Printing output
   cout << "Triplets adds to target sum "<< targetsum << " are: " << endl;
   for (const auto i: out) {
      for (const auto j: i) {
         cout << j << " ";
      }
      cout << endl;
   }
}
